File: tutorials/go/debugging/ast/custom_linter.go
package main

import (
	"flag"
	"fmt"
	"go/ast"
	"go/parser"
	"go/token"
	"os"
	"path/filepath"
	"strings"
)

// Issue represents a linting issue
type Issue struct {
	Pos      token.Position
	Message  string
	Severity string
}

// LintRule defines a rule for linting
type LintRule interface {
	Check(fset *token.FileSet, file *ast.File) []Issue
	Name() string
	Description() string
}

// UnusedImportRule checks for unused imports
type UnusedImportRule struct{}

func (r UnusedImportRule) Name() string {
	return "unused-import"
}

func (r UnusedImportRule) Description() string {
	return "Detects unused imports in the code"
}

func (r UnusedImportRule) Check(fset *token.FileSet, file *ast.File) []Issue {
	var issues []Issue
	
	// Get all imports
	imports := make(map[string]token.Position)
	for _, imp := range file.Imports {
		name := ""
		if imp.Name != nil {
			// Named import
			name = imp.Name.Name
			if name == "_" {
				// Blank import is used for side effects
				continue
			}
			if name == "." {
				// Dot import is hard to track, skip for now
				continue
			}
		} else {
			// Regular import, extract the package name
			path := strings.Trim(imp.Path.Value, "\"")
			parts := strings.Split(path, "/")
			name = parts[len(parts)-1]
		}
		imports[name] = fset.Position(imp.Pos())
	}
	
	// Find all identifiers in the file
	ast.Inspect(file, func(n ast.Node) bool {
		if ident, ok := n.(*ast.Ident); ok {
			// Skip checking package selectors (e.g., fmt.Printf)
			if _, ok := imports[ident.Name]; ok {
				// This identifier matches an import name, remove it from the map
				delete(imports, ident.Name)
			}
		}
		
		// For SelectorExpr (e.g., fmt.Printf), check the package part
		if sel, ok := n.(*ast.SelectorExpr); ok {
			if x, ok := sel.X.(*ast.Ident); ok {
				if _, ok := imports[x.Name]; ok {
					// Used import, remove it from the map
					delete(imports, x.Name)
				}
			}
		}
		
		return true
	})
	
	// Remaining imports in the map are unused
	for name, pos := range imports {
		issues = append(issues, Issue{
			Pos:      pos,
			Message:  fmt.Sprintf("Unused import: %s", name),
			Severity: "warning",
		})
	}
	
	return issues
}

// ErrorReturnRule checks if errors are being checked
type ErrorReturnRule struct{}

func (r ErrorReturnRule) Name() string {
	return "error-check"
}

func (r ErrorReturnRule) Description() string {
	return "Ensures that errors returned from function calls are checked"
}

func (r ErrorReturnRule) Check(fset *token.FileSet, file *ast.File) []Issue {
	var issues []Issue
	
	// Find all assignments
	ast.Inspect(file, func(n ast.Node) bool {
		switch stmt := n.(type) {
		case *ast.AssignStmt:
			// Check if right side is a function call that might return an error
			for _, rhs := range stmt.Rhs {
				if call, ok := rhs.(*ast.CallExpr); ok {
					// If the assignment has multiple left-hand values and more than one right-hand value,
					// we need to check if the last one might be an error
					if len(stmt.Lhs) > 1 && len(stmt.Rhs) == 1 {
						// Check if the last left-hand value is being assigned to _
						if len(stmt.Lhs) >= 2 {
							lastLhs := stmt.Lhs[len(stmt.Lhs)-1]
							if ident, ok := lastLhs.(*ast.Ident); ok && ident.Name == "_" {
								// Error is being explicitly ignored
								issues = append(issues, Issue{
									Pos:      fset.Position(ident.Pos()),
									Message:  "Error is explicitly ignored with _",
									Severity: "warning",
								})
							}
						}
					}
				}
			}
		case *ast.ExprStmt:
			// Check for function calls whose return values are completely ignored
			if call, ok := stmt.X.(*ast.CallExpr); ok {
				// Try to determine if the function might return an error
				// This is a simplistic approach - in a real linter, we would use type information
				if funcName, ok := getFunctionName(call); ok {
					if strings.HasPrefix(funcName, "Create") ||
					   strings.HasPrefix(funcName, "New") ||
					   strings.HasPrefix(funcName, "Open") ||
					   strings.HasPrefix(funcName, "Read") ||
					   strings.HasPrefix(funcName, "Write") {
						issues = append(issues, Issue{
							Pos:      fset.Position(call.Pos()),
							Message:  fmt.Sprintf("Result of %s is ignored, but it might return an error", funcName),
							Severity: "warning",
						})
					}
				}
			}
		}
		return true
	})
	
	return issues
}

// Helper function to get the function name from a CallExpr
func getFunctionName(call *ast.CallExpr) (string, bool) {
	switch fun := call.Fun.(type) {
	case *ast.Ident:
		// Direct function call, e.g., doSomething()
		return fun.Name, true
	case *ast.SelectorExpr:
		// Package or method call, e.g., pkg.Function() or obj.Method()
		if ident, ok := fun.X.(*ast.Ident); ok {
			return ident.Name + "." + fun.Sel.Name, true
		}
		return fun.Sel.Name, true
	default:
		return "", false
	}
}

// Run the linter on a file
func lintFile(filename string, rules []LintRule) ([]Issue, error) {
	// Create file set for position information
	fset := token.NewFileSet()
	
	// Parse the file
	file, err := parser.ParseFile(fset, filename, nil, parser.ParseComments)
	if err != nil {
		return nil, fmt.Errorf("failed to parse file: %v", err)
	}
	
	// Apply all rules
	var allIssues []Issue
	for _, rule := range rules {
		issues := rule.Check(fset, file)
		allIssues = append(allIssues, issues...)
	}
	
	return allIssues, nil
}

func main() {
	// Parse command line flags
	dir := flag.String("dir", ".", "Directory to lint")
	recursive := flag.Bool("recursive", false, "Recursively lint subdirectories")
	flag.Parse()
	
	// Create lint rules
	rules := []LintRule{
		UnusedImportRule{},
		ErrorReturnRule{},
	}
	
	// Process files
	var filesToLint []string
	
	walkFn := func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() {
			if !*recursive && path != *dir {
				return filepath.SkipDir
			}
			return nil
		}
		if strings.HasSuffix(path, ".go") && !strings.HasSuffix(path, "_test.go") {
			filesToLint = append(filesToLint, path)
		}
		return nil
	}
	
	if err := filepath.Walk(*dir, walkFn); err != nil {
		fmt.Fprintf(os.Stderr, "Error walking directory: %v\n", err)
		os.Exit(1)
	}
	
	// Lint each file
	issueCount := 0
	for _, file := range filesToLint {
		issues, err := lintFile(file, rules)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error linting %s: %v\n", file, err)
			continue
		}
		
		for _, issue := range issues {
			fmt.Printf("%s:%d:%d: %s: %s\n",
				issue.Pos.Filename,
				issue.Pos.Line,
				issue.Pos.Column,
				issue.Severity,
				issue.Message)
			issueCount++
		}
	}
	
	// Print summary
	fmt.Printf("\nLinted %d files, found %d issues\n", len(filesToLint), issueCount)
	
	// Return non-zero exit code if issues were found
	if issueCount > 0 {
		os.Exit(1)
	}
}

File: tutorials/rust/debugging/ast/custom_linter.rs
use syn::{parse_file, Item, ItemFn, Expr, ExprMatch, Pat, Arm};
use syn::visit::{self, Visit};
use std::fs::File;
use std::io::Read;
use std::path::Path;

// Linting rule struct
struct LintRule {
    name: String,
    description: String,
    check_fn: fn(&syn::File) -> Vec<LintError>,
}

// Error reported by a lint rule
struct LintError {
    message: String,
    line: usize,
    column: usize,
    rule_name: String,
}

impl LintError {
    fn new(message: String, rule_name: &str) -> Self {
        LintError {
            message,
            line: 0,  // Would be populated from span in a real implementation
            column: 0, // Would be populated from span in a real implementation
            rule_name: rule_name.to_string(),
        }
    }
}

// Custom linter implementation
struct RustLinter {
    rules: Vec<LintRule>,
}

impl RustLinter {
    fn new() -> Self {
        // Create a linter with default rules
        let mut linter = RustLinter {
            rules: Vec::new(),
        };
        
        // Add default rules
        linter.add_rule(LintRule {
            name: "exhaustive_match".to_string(),
            description: "Checks for non-exhaustive match expressions".to_string(),
            check_fn: check_exhaustive_match,
        });
        
        linter.add_rule(LintRule {
            name: "unwrap_used".to_string(),
            description: "Detects usage of unwrap() which might panic".to_string(),
            check_fn: check_unwrap_usage,
        });
        
        linter.add_rule(LintRule {
            name: "complex_function".to_string(),
            description: "Identifies functions that are too large or complex".to_string(),
            check_fn: check_complex_functions,
        });
        
        linter
    }
    
    fn add_rule(&mut self, rule: LintRule) {
        self.rules.push(rule);
    }
    
    fn lint_file(&self, path: &Path) -> Result<Vec<LintError>, Box<dyn std::error::Error>> {
        // Read the file
        let mut file = File::open(path)?;
        let mut content = String::new();
        file.read_to_string(&mut content)?;
        
        // Parse the file
        let syntax = parse_file(&content)?;
        
        // Apply all rules
        let mut errors = Vec::new();
        for rule in &self.rules {
            let rule_errors = (rule.check_fn)(&syntax);
            errors.extend(rule_errors);
        }
        
        Ok(errors)
    }
}

// Rule implementation: Check for non-exhaustive match expressions
fn check_exhaustive_match(file: &syn::File) -> Vec<LintError> {
    struct MatchVisitor {
        errors: Vec<LintError>,
    }
    
    impl<'ast> Visit<'ast> for MatchVisitor {
        fn visit_expr_match(&mut self, node: &'ast ExprMatch) {
            // Check if the match has a wildcard pattern
            let has_wildcard = node.arms.iter().any(|arm| {
                matches!(arm.pat, Pat::Wild(_))
            });
            
            // If it doesn't have a wildcard, it might not be exhaustive
            if !has_wildcard {
                self.errors.push(LintError::new(
                    "Match expression might not be exhaustive. Consider adding a wildcard '_' pattern".to_string(),
                    "exhaustive_match"
                ));
            }
            
            // Continue visiting
            visit::visit_expr_match(self, node);
        }
    }
    
    let mut visitor = MatchVisitor { errors: Vec::new() };
    visitor.visit_file(file);
    visitor.errors
}

// Rule implementation: Check for unwrap() usage
fn check_unwrap_usage(file: &syn::File) -> Vec<LintError> {
    struct UnwrapVisitor {
        errors: Vec<LintError>,
    }
    
    impl<'ast> Visit<'ast> for UnwrapVisitor {
        fn visit_expr(&mut self, node: &'ast Expr) {
            // Look for method calls
            if let Expr::MethodCall(method_call) = node {
                // Check if method name is unwrap
                if method_call.method == "unwrap" {
                    self.errors.push(LintError::new(
                        "Use of unwrap() detected. Consider using ? or match/if let for error handling".to_string(),
                        "unwrap_used"
                    ));
                }
            }
            
            // Continue visiting
            visit::visit_expr(self, node);
        }
    }
    
    let mut visitor = UnwrapVisitor { errors: Vec::new() };
    visitor.visit_file(file);
    visitor.errors
}

// Rule implementation: Check for complex functions
fn check_complex_functions(file: &syn::File) -> Vec<LintError> {
    struct ComplexityVisitor {
        errors: Vec<LintError>,
    }
    
    impl<'ast> Visit<'ast> for ComplexityVisitor {
        fn visit_item_fn(&mut self, node: &'ast ItemFn) {
            // Simple complexity metric: count statements
            let stmt_count = node.block.stmts.len();
            
            // If function has too many statements, report it
            if stmt_count > 20 {  // Arbitrary threshold
                self.errors.push(LintError::new(
                    format!("Function '{}' has {} statements, which exceeds the recommended maximum of 20", 
                            node.sig.ident, stmt_count),
                    "complex_function"
                ));
            }
            
            // Continue visiting
            visit::visit_item_fn(self, node);
        }
    }
    
    let mut visitor = ComplexityVisitor { errors: Vec::new() };
    visitor.visit_file(file);
    visitor.errors
}

fn main() {
    // Example Rust code to lint
    let example_code = r#"
    fn process_data(data: Option<String>) -> String {
        // Using unwrap which might panic
        let value = data.unwrap();
        
        // Non-exhaustive match
        let result = match value.as_str() {
            "hello" => "world",
            "goodbye" => "friend",
            // Missing wildcard case
        };
        
        result.to_string()
    }
    
    fn very_complex_function() {
        let mut sum = 0;
        
        // Lots of statements to trigger complexity warning
        for i in 0..100 {
            sum += i;
        }
        println!("Step 1");
        println!("Step 2");
        println!("Step 3");
        println!("Step 4");
        println!("Step 5");
        println!("Step 6");
        println!("Step 7");
        println!("Step 8");
        println!("Step 9");
        println!("Step 10");
        println!("Step 11");
        println!("Step 12");
        println!("Step 13");
        println!("Step 14");
        println!("Step 15");
        println!("Final sum: {}", sum);
    }
    "#;
    
    // Write to a temporary file for linting
    let temp_file = "temp_lint_example.rs";
    std::fs::write(temp_file, example_code).expect("Failed to write temporary file");
    
    // Create a linter and run it
    let linter = RustLinter::new();
    let path = Path::new(temp_file);
    
    match linter.lint_file(path) {
        Ok(errors) => {
            println!("Found {} lint issues:", errors.len());
            for error in errors {
                println!("[{}] {}", error.rule_name, error.message);
            }
        },
        Err(e) => eprintln!("Error linting file: {}", e),
    }
    
    // Clean up
    std::fs::remove_file(temp_file).expect("Failed to remove temporary file");
}
